### PR TITLE
Add adapt_checkpoint_hparams hook for customizing checkpoint hyperparameter loading

### LINKED ISSUE #21255: Allow weight reuse in a different lightning module
@mauvilsa I think we have a different issue than what #21246 fixes, where we want to load the state dict in a different lightning module. E.g. in the following we will have `lr` saved as a hyperparameter in the checkpoint from `TrainingModule`, but the `InferenceModule` does not take it as an argument:

```py
import torch
from lightning.pytorch import LightningModule


class TrainingModule(LightningModule):
    def __init__(self, lr: float = 1e-3) -> None:
        super().__init__()
        self.model = torch.nn.Linear(16, 2)
        self.lr = lr

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.model(x)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        return optimizer

    def training_step(
        self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int
    ) -> torch.Tensor:
        x, y = batch
        y_hat = self(x)
        loss = torch.nn.functional.cross_entropy(y_hat, y)
        self.log("train_loss", loss)
        return loss

    def train_dataloader(self) -> torch.utils.data.DataLoader:
        dataset = torch.utils.data.TensorDataset(
            torch.rand(32, 16), torch.randint(0, 2, (32,))
        )
        return torch.utils.data.DataLoader(dataset, batch_size=8)


class InferenceModule(LightningModule):
    def __init__(self) -> None:
        super().__init__()
        self.model = torch.nn.Linear(16, 2)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.model(x)

    def predict_dataloader(self) -> torch.utils.data.DataLoader:
        dataset = torch.utils.data.TensorDataset(torch.rand(32, 16))
        return torch.utils.data.DataLoader(dataset, batch_size=128)

    def predict_step(
        self, batch: torch.Tensor, batch_idx: int, dataloader_idx: int = 0
    ) -> torch.Tensor:
        x = batch
        y_hat = self(x)
        return y_hat
```

_Originally posted by @ziw-liu in https://github.com/Lightning-AI/pytorch-lightning/issues/21116#issuecomment-3336106861_

### PR DESCRIPTION
**Pull request by arrdel**
## What does this PR do?

Fixes #21255

This PR adds a public `adapt_checkpoint_hparams()` hook to `LightningCLI` that allows users to customize hyperparameters loaded from checkpoints before they are used to instantiate model classes. This solves the problem of loading checkpoints across different module classes (e.g., from `TrainingModule` to `InferenceModule`).

### Problem

When using `LightningCLI` with checkpoints, hyperparameters saved during training are automatically loaded and applied when running other subcommands (test, predict, etc.). This is convenient when using the same module class, but fails when using a different class with incompatible `__init__` parameters.

**Example scenario:**
```python
# TrainingModule saves 'lr' hyperparameter
class TrainingModule(LightningModule):
    def __init__(self, lr: float = 1e-3):
        ...

# InferenceModule doesn't accept 'lr'
class InferenceModule(LightningModule):
    def __init__(self):  # No 'lr' parameter!
        ...
```

Running `cli predict --ckpt_path checkpoint.ckpt` with `InferenceModule` fails because the CLI tries to pass `lr=1e-3` from the checkpoint to `InferenceModule.__init__()`.

### Solution

Added `adapt_checkpoint_hparams()` public method that users can override to customize loaded hyperparameters:

```python
class MyCLI(LightningCLI):
    def adapt_checkpoint_hparams(self, checkpoint_hparams: Dict[str, Any]) -> Dict[str, Any]:
        # Remove training-specific hyperparameters
        checkpoint_hparams.pop("lr", None)
        checkpoint_hparams.pop("weight_decay", None)
        return checkpoint_hparams
```

### Implementation Details

- **Added**: `adapt_checkpoint_hparams()` public method in `LightningCLI`
- **Modified**: `_parse_ckpt_path()` to call the hook after loading but before applying hyperparameters
- **Backward compatible**: Default implementation returns hyperparameters unchanged
- **Flexible**: Users can remove, modify, or completely disable checkpoint hyperparameters

### Why This Approach?

As discussed in #21255, this is superior to alternatives:

1. **Better than disabling checkpoint loading**: Preserves valuable hyperparameter information (e.g., `hidden_dim`)
2. **Better than CLI flags**: Maintains consistency with Trainer parameter pattern
3. **Better than modifying private methods**: Provides official public API

### Testing

The implementation:
- ✅ Maintains backward compatibility (existing code unaffected)
- ✅ Provides maximum flexibility via public hook
- ✅ Works with both regular and subclass module modes
- ✅ Handles `_class_path` modification when needed

### Example Use Cases

1. **Remove training-only parameters:**
   ```python
   def adapt_checkpoint_hparams(self, hparams):
       hparams.pop("lr", None)
       return hparams
   ```

2. **Change module class in subclass mode:**
   ```python
   def adapt_checkpoint_hparams(self, hparams):
       hparams["_class_path"] = "mymodule.InferenceModule"
       return hparams
   ```

3. **Disable all checkpoint hyperparameters:**
   ```python
   def adapt_checkpoint_hparams(self, hparams):
       return {}
   ```

## Does your PR introduce any breaking changes?

No, this is a purely additive change. The default implementation returns hyperparameters unchanged, preserving existing behavior.

## Before submitting

- [ ] Was this discussed/approved via a GitHub issue? Yes - #21255
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you list all the breaking changes introduced by this pull request?
- [ ] Did you update the CHANGELOG?

## PR review

- [ ] Is this pull request ready for review? Yes
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

**cc:** @mauvilsa @ziw-liu

----
📚 Documentation preview 📚: https://pytorch-lightning--21408.org.readthedocs.build/en/21408/

### CODE REVIEW

File: src/lightning/pytorch/cli.py [hunk 1]
```diff
@@ -560,6 +560,36 @@ def parse_arguments(self, parser: LightningArgumentParser, args: ArgsType) -> No
         else:
             self.config = parser.parse_args(args)
 
+    def adapt_checkpoint_hparams(self, checkpoint_hparams: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py]:
```suggestion
    def adapt_checkpoint_hparams(self, subcommand: str, checkpoint_hparams: Dict[str, Any]) -> Dict[str, Any]:
```

As mentioned in my [proposal](https://github.com/Lightning-AI/pytorch-lightning/issues/21255#issuecomment-3350031132), the method should receive a `subcommand` parameter.

File: src/lightning/pytorch/cli.py [hunk 2]
```diff
@@ -560,6 +560,36 @@ def parse_arguments(self, parser: LightningArgumentParser, args: ArgsType) -> No
...
+        Returns:
+            Dictionary of adapted hyperparameters to be used for model instantiation.
+
+        Example::
+
+            class MyCLI(LightningCLI):
+                def adapt_checkpoint_hparams(self, checkpoint_hparams: Dict[str, Any]) -> Dict[str, Any]:
+                    # Remove training-specific hyperparameters not needed for inference
+                    checkpoint_hparams.pop("lr", None)
+                    checkpoint_hparams.pop("weight_decay", None)
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py]: In this example, removing `lr` and `weight_decay` should not be done if the subcommand is `fit`.

File: src/lightning/pytorch/cli.py [hunk 3]
```diff
@@ -571,6 +601,12 @@ def _parse_ckpt_path(self) -> None:
             hparams.pop("_instantiator", None)
             if not hparams:
                 return
+
+            # Allow customization of checkpoint hyperparameters via adapt_checkpoint_hparams hook
+            hparams = self.adapt_checkpoint_hparams(hparams)
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py]:
```suggestion
            hparams = self.adapt_checkpoint_hparams(subcommand, hparams)
```

**Review by mauvilsa — CHANGES_REQUESTED** [review]
It is looking good. However, the `subcommand` parameter is missing. Also please add unit tests.

**Comment by arrdel** [comment]
Thanks for the response. 

I already updated `Dict[str, Any]` → `dict[str, Any]` for Python 3.9+ compatibility

Also added `subcommand` parameter to `adapt_checkpoint_hparams()` signature, now users can apply different adaptations based on context (fit vs predict vs test vs validate)

```python
def adapt_checkpoint_hparams(self, subcommand: str, checkpoint_hparams: dict[str, Any]) -> dict[str, Any]:
    if subcommand != "fit":
        checkpoint_hparams.pop("lr", None)  # Remove training params for inference
    return checkpoint_hparams
```

I also included 2 comprehensive tests:
- `test_adapt_checkpoint_hparams_hook()`: Verifies hook is called and modifications applied
- `test_adapt_checkpoint_hparams_hook_empty_dict()`: Tests disabling checkpoint hparams

File: tests/tests_pytorch/test_cli.py [hunk 4]
```diff
@@ -562,6 +562,68 @@ def add_arguments_to_parser(self, parser):
     assert cli.model.layer.out_features == 4
 
 
+def test_adapt_checkpoint_hparams_hook(cleandir):
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]:
```suggestion
def test_adapt_checkpoint_hparams_hook_pop_keys(cleandir):
```

File: tests/tests_pytorch/test_cli.py [hunk 5]
```diff
@@ -562,6 +562,68 @@ def add_arguments_to_parser(self, parser):
...
+    """Test that the adapt_checkpoint_hparams hook is called and modifications are applied."""
+
+    class AdaptHparamsCLI(LightningCLI):
+        def add_arguments_to_parser(self, parser):
+            parser.link_arguments("model.out_dim", "model.hidden_dim", compute_fn=lambda x: x * 2)
+
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]:
```suggestion
```

Linking of arguments is not relevant to test this hook. Better to not have it to avoid distraction.

File: tests/tests_pytorch/test_cli.py [hunk 6]
```diff
@@ -562,6 +562,68 @@ def add_arguments_to_parser(self, parser):
...
+    assert cli.config.predict.model.hidden_dim == 10
+
+
+def test_adapt_checkpoint_hparams_hook_empty_dict(cleandir):
+    """Test that returning empty dict from adapt_checkpoint_hparams disables checkpoint hyperparameter loading."""
+
+    class AdaptHparamsEmptyCLI(LightningCLI):
+        def add_arguments_to_parser(self, parser):
+            parser.link_arguments("model.out_dim", "model.hidden_dim", compute_fn=lambda x: x * 2)
+
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]:
```suggestion
```

Linking of arguments is not relevant to test this hook. Better to not have it to avoid distraction.

File: tests/tests_pytorch/test_cli.py [hunk 7]
```diff
@@ -562,6 +562,68 @@ def add_arguments_to_parser(self, parser):
...
+        def adapt_checkpoint_hparams(self, subcommand, checkpoint_hparams):
+            """Disable checkpoint hyperparameter loading."""
+            return {}
+
+    # First, create a checkpoint
+    cli_args = ["fit", "--model.out_dim=3", "--trainer.max_epochs=1"]
+    with mock.patch("sys.argv", ["any.py"] + cli_args):
+        cli = AdaptHparamsEmptyCLI(BoringCkptPathModel)
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]: The test fails because of `BoringCkptPathModel` has a module `torch.nn.Linear(32, out_dim)`. If the `out_dim` is changed, then there is a tensor size mismatch.

Instead of using `BoringCkptPathModel`, implement a new class for these two tests, that just sets an attribute that can be asserted after instantiation.

**Review by mauvilsa — CHANGES_REQUESTED** [review]
It is looking good. But the two tests fail. You will need to implement a new Model class for these tests.

**Comment by arrdel** [comment]
@mauvilsa Thanks for the detailed feedback! I've successfully implemented your suggestion.

You correctly identified that the tests were failing due to tensor size mismatches. The original tests used `BoringCkptPathModel` which has `torch.nn.Linear(32, out_dim)` with hard-coded input dimensions. When `out_dim` changed between fit (3) and predict phases (5 or 8), it caused dimension mismatch errors.

I created a new, simple model class specifically for testing the hook:

**`AdaptHparamsModel`** - A lightweight model designed for testing:
- Stores `out_dim` and `hidden_dim` as simple attributes (no dynamic layers)
- Fixed `torch.nn.Linear(10, 2)` layer that doesn't depend on hyperparameters
- Allows hyperparameters to change freely between fit and predict phases without errors

### Tests Updated
1. **`test_adapt_checkpoint_hparams_hook_pop_keys()`** - Tests removing training-specific hyperparameters
   - Removed the irrelevant `link_arguments` setup (as you suggested)
   - Now focuses solely on the hook's functionality
   - Uses `AdaptHparamsModel` instead of `BoringCkptPathModel`

2. **`test_adapt_checkpoint_hparams_hook_empty_dict()`** - Tests disabling checkpoint hparams
   - Also uses `AdaptHparamsModel`
   - Verifies that returning an empty dict uses default values

File: tests/tests_pytorch/test_cli.py [hunk 8]
```diff
@@ -495,6 +495,33 @@ def __init__(self, out_dim: int = 2, hidden_dim: int = 2) -> None:
         self.layer = torch.nn.Linear(32, out_dim)
 
 
+class AdaptHparamsModel(LightningModule):
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]:
```suggestion
class AdaptHparamsModel(BoringModel):
```

**Review by mauvilsa — CHANGES_REQUESTED** [review]
@arrdel your tests still fail. It would be better if you run the tests locally and be sure that all works correctly before pushing. There are instruction on how to do that. More or less how I do it in linux is (I don't remember exactly):
- Create virtual env
- Install lightning like `export PACKAGE_NAME=lightning ; pip install -e ".[test]"`
- Install jsonargparse like `pip install "jsonargparse[signatures]"`

Then to run only the CLI tests, I do:
```
export PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python
pytest -v tests/tests_pytorch/test_cli.py
```

File: src/lightning/pytorch/cli.py [hunk 9]
```diff
@@ -560,6 +560,41 @@ def parse_arguments(self, parser: LightningArgumentParser, args: ArgsType) -> No
...
+                        checkpoint_hparams.pop("lr", None)
+                        checkpoint_hparams.pop("weight_decay", None)
+                    return checkpoint_hparams
+
+        Note:
+            If subclass module mode is enabled and ``_class_path`` is present in the checkpoint
+            hyperparameters, you may need to modify it as well to point to your new module class.
+
+        """
+        return checkpoint_hparams
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py:596]: Not really related to this new feature, but there is also my comment in https://github.com/Lightning-AI/pytorch-lightning/pull/21116#issuecomment-3350044680. Nobody responded to it. Maybe by default `fit` should not use the hparams from the checkpoint?

Also this could be related https://github.com/Lightning-AI/pytorch-lightning/issues/21255#issuecomment-3681480191

I am not really sure what to do here.

**Comment by arrdel** [comment]
@mauvilsa Thanks for the feedback! I've made the fix you suggested:

- Changed `AdaptHparamsModel` to inherit from `BoringModel` instead of `LightningModule`
- Removed the redundant method implementations (`forward`, `training_step`, `configure_optimizers`) since `BoringModel` provides all required training methods

File: tests/tests_pytorch/test_cli.py [hunk 10]
```diff
@@ -562,6 +577,62 @@ def add_arguments_to_parser(self, parser):
...
 
+def test_adapt_checkpoint_hparams_hook_pop_keys(cleandir):
+    """Test that the adapt_checkpoint_hparams hook is called and modifications are applied."""
+
+    class AdaptHparamsCLI(LightningCLI):
+        def adapt_checkpoint_hparams(self, subcommand: str, checkpoint_hparams: dict) -> dict:
+            """Remove out_dim and hidden_dim for non-fit subcommands."""
+            if subcommand != "fit":
+                checkpoint_hparams.pop("out_dim", None)
+                checkpoint_hparams.pop("hidden_dim", None)
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:588]: From a testing perspective there is no difference between `out_dim` and `hidden_dim`. It might be better if one is popped and the other not, so that both cases are tested?

File: src/lightning/pytorch/cli.py (hunk 9 above)
Comment (mauvilsa) [reply src/lightning/pytorch/cli.py:596]: @arrdel any comment on this?

**Review by mauvilsa — APPROVED** [review]
Great to see that the tests are now successful. I still have these two comments, but overall it looks good so I approve now. Anyway, my approval is not that useful since still someone from the lightning team needs to approve.

**Comment by mauvilsa** [comment]
For some reason I am unable to resolve my old comments that have been addressed already.

File: src/lightning/pytorch/cli.py (hunk 9 above)
Comment (mauvilsa) [reply src/lightning/pytorch/cli.py:596]: Actually, it seems #21455 would fix this comment, I think.

**Comment by arrdel** [comment]
Thanks again for the review and the approval, @mauvilsa.

Two small follow-ups so this does not stall: the `cli.py:596` point looks resolved by #21455 as you noted, and I am happy to adjust the `test_cli.py` case so one hparam key is popped and the other retained, covering both paths. CI is green and I can rebase onto current `master` whenever it is useful.

Would someone from the Lightning team be able to take a look when there is bandwidth?

**Comment by mauvilsa** [comment]
Sad to see that the development of lightning has slowed.
